File: 465_Optimal_Account_Balancing.cpp
class Solution {
private:
    vector<long> dept;
    int dfs(vector<long>& dept, int s) {
        while (s < dept.size() && !dept[s]) s++;
        int res = INT_MAX;
        long prev = 0;
        for (int i = s + 1; i < dept.size(); ++i) {
            if (dept[i] != prev && dept[s] * dept[i] < 0) {
                dept[i] += dept[s];
                res = min(res, 1 + dfs(dept, s + 1));
                prev = dept[i] -= dept[s];
            }
        }
        return res == INT_MAX ? 0 : res;
    }
public:
    int minTransfers(vector<vector<int>>& transactions) {
        vector<long> dept;
        unordered_map<int, long> balance;
        for (auto t : transactions) {
            balance[t[0]] -= t[2];
            balance[t[1]] += t[2];
        }
        for (auto b : balance) {
            if (b.second != 0) dept.push_back(b.second);
        }
        return dfs(dept, 0);
    }
};

File: 465_Optimal_Account_Balancing.py
class Solution:
    def minTransfers(self, transactions: List[List[int]]) -> int:
        balances = collections.defaultdict(int)
        for source, target, amount in transactions:
            balances[source] -= amount
            balances[target] += amount

        balances_clear = collections.defaultdict(int)
        base = 0
        for v in balances.values():
            if v == 0:
                continue
            if -v in balances_clear:
                balances_clear[-v] -= 1
                base += 1
            else:
                balances_clear[v] += 1
                
        net_profit = []
        for key, val in balances_clear.items():
            net_profit.extend([key] * val)
        
        n = len(net_profit)
        
        def dfs(i):
            if i >= n - 1:
                return 0
            if net_profit[i] == 0:
                return dfs(i + 1)
            
            p = net_profit[i]
            res = math.inf
            seen = set()
            for j in range(i + 1, n):
                if net_profit[j] * p < 0 and net_profit[j] not in seen:
                    net_profit[j] += p
                    res = min(res, 1 + dfs(i + 1))
                    net_profit[j] -= p
                    seen.add(net_profit[j])
                    if net_profit[i] + net_profit[j] == 0:
                        return res
            return res
        
        return dfs(0) + base